compiler/src/org.graalvm.compiler.loop.phases/src/org/graalvm/compiler/loop/phases/LoopPredicationPhase.java

```diff
@@ -0,0 +1,399 @@
+        if (graph.hasLoops() && graph.getGuardsStage().allowsFloatingGuards() && context.getOptimisticOptimizations().useLoopLimitChecks(graph.getOptions()) && speculationLog != null) {
+            LoopsData data = new LoopsData(graph);
+            data.detectedCountedLoops();
+            try (DebugContext.Scope s = debug.scope("predication", data.getCFG())) {
+                for (LoopEx loop : data.countedLoops()) {
+                    final FrameState state = loop.loopBegin().stateAfter();
+                    final BytecodePosition pos = new BytecodePosition(null, state.getMethod(), state.bci);
+                    SpeculationLog.SpeculationReason reason = LOOP_PREDICATION.createSpeculationReason(pos);
+                    if (speculationLog.maySpeculate(reason)) {
+                        final CountedLoopInfo counted = loop.counted();
+                        final InductionVariable counter = counted.getCounter();
+                        final Condition condition = ((CompareNode) counted.getLimitTest().condition()).condition().asCondition();
+                        // Only inner most loops with int induction variable
+                        if (loop.loop().getChildren().isEmpty() &&
+                                        (((IntegerStamp) counter.valueNode().stamp(NodeView.DEFAULT)).getBits() == 32) &&
+                                        !counted.isUnsignedCheck() &&
+                                        ((condition != NE && condition != EQ) || (counter.isConstantStride() && Math.abs(counter.constantStride()) == 1)) &&
+                                        (loop.loopBegin().isMainLoop() || loop.loopBegin().isSimpleLoop())) {
+
+                            prepareGuards(loop);
+
+                            // C2 only applies loop predication to guards dominating the backedge.
+                            // The following logic emulates that behavior.
+                            NodeIterable<GuardNode> guards = loop.whole().nodes().filter(GuardNode.class);
+                            if (LoopPredicationMainPath.getValue(graph.getOptions())) {
+                                EconomicSet<Node> mainPathNodes = collectMainPath(data, loop);
+                                guards = guards.filter(n -> mainPathNodes.contains(n));
+                            }
+                            for (GuardNode guard : guards) {
+                                if (guard.usages().filter(n -> !(n instanceof GuardedNode)).count() > 0) {
+                                    continue;
+                                }
+                                processGuard(loop, guard);
+                            }
+                        }
+                    }
+                }
+            } catch (Throwable t) {
+                throw debug.handle(t);
+            }
+        }
+    }
+
+    private static EconomicSet<Node> collectMainPath(LoopsData data, LoopEx loop) {
```

Comment:
Could you clarify what this is about in a comment? Is it about avoiding paths that are conceptually loop exits but have not yet been marked as such in the computation of `loop.whole()`?

Reply:
In C2 guards do not float. There are pinned in whatever cfg path they were found at parse time. The C2 optimization pass that this PR emulates looks for candidate guards by starting from the loop backedge and following the dominator chain to the loop head. So any guard that's pinned in a path that's not always executed is not a candidate for predication. This code emulates that behavior (or at least it tries to).

I'm not sure what's the rationale for the C2 implementation is. Maybe it was simpler to do it that way and it has work well enough that it wasn't revised for a long time. I could see how guards that are not always executed could sometimes fail, cause the whole loop predication optimization to be skipped on recompilation and result in a performance anomaly.

The reason I include this conservative heuristic is that it feels safer from a performance point of view.

Reply:
OK, thanks for the details

Reply:
Should this also be limited to guards that are anchored below the counter check of the loop?